platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,890 @@
+                    // but it allows us to use `&mut [u8]` in the DescriptorBuilder.
+                    let slice = unsafe { core::slice::from_raw_parts_mut(buff_addr, buff_size) };
+                    remaining = remaining.saturating_sub(DESCR_BUFF_SIZE);
+                    let first = index == 0;
+                    let last = remaining == 0;
+                    descriptors[index] = idmac::DescriptorBuilder::new()
+                        .disable_irq(!last)
+                        .first(first)
+                        .last(last)
+                        .link((!last).then(|| (&descriptors[index + 1]).into()))
+                        .expect("Should be able to link to next descriptor")
+                        .buff_slice(slice)
+                        .map_err(|_| sdmmc::Error::Buffer)?
+                        .build();
+                    index += 1;
+                }
+
+                // TODO: should this return some kind of guard, which needs to be used later
+                // to make sure the descriptor has a long enough lifetime?
+                unsafe { self.prepare_dma(&descriptors[0], *cnt as u32) };
+            }
+            _ => (),
+        }
+
+        // Set the argument to be sent
+        self.smhc
+            .smhc_cmdarg
+            .write(|w| unsafe { w.bits(params.argument) });
+
+        // Configure and start command
+        self.smhc.smhc_cmd.write(|w| {
+            w.cmd_load().set_bit();
+            w.wait_pre_over().wait();
+            w.stop_cmd_flag().bit(auto_stop);
+            w.data_trans().bit(data_trans);
+            w.trans_dir().bit(trans_dir);
+            w.chk_resp_crc().bit(chk_resp_crc);
+            w.long_resp().bit(long_resp);
+            w.resp_rcv().bit(resp_rcv);
+            w.cmd_idx().variant(cmd_idx);
+            w
+        });
+
+        guard.data.state = State::WaitForCommand;
+        // Now wait for completion or error interrupt
+        guard.wait_for_irq().await;
+        tracing::trace!("SMHC operation completed");
+        if let Some(error) = guard.data.err.take() {
+            tracing::warn!(?error, "SMHC error");
+            Err(match error {
+                ErrorKind::Response => sdmmc::Error::Response,
+                ErrorKind::ResponseCrc => sdmmc::Error::Response,
+                ErrorKind::DataCrc => sdmmc::Error::Data,
+                ErrorKind::ResponseTimeout => sdmmc::Error::Timeout,
+                ErrorKind::DataTimeout => sdmmc::Error::Timeout,
+                ErrorKind::DataStarvationTimeout => sdmmc::Error::Timeout,
+                ErrorKind::FifoUnderrunOverflow => sdmmc::Error::Data,
+                ErrorKind::CommandBusyIllegalWrite => sdmmc::Error::Busy,
+                ErrorKind::DataStart => sdmmc::Error::Data,
+                ErrorKind::DataEnd => sdmmc::Error::Data,
+                ErrorKind::Dma => sdmmc::Error::Other,
+                ErrorKind::Other => sdmmc::Error::Other,
```

Comment:
hmm, it's a bit unfortunate that some potentially useful information is lost here when converting these `ErrorKind`s into `sdmmc::Error`s. for example, the user cannot determine whether a `sdmmc::Error::Timeout` represents a `ResponseTimeout`, a `DataTimeout`, or a `DataStarvationTimeout`.

what do you think about changing the `sdmmc::Error` type to something like this:
```rust
pub struct Error {
    kind: sdmmc::ErrorKind,
    message: Option<&'static str>, 
}
```
(where `sdmmc::ErrorKind` is the enum currently named `sdmmc::Error`).

this way, users can still programmatically match on the enum, representing a standardized enumeration of possible kinds of errors that a SDMMC driver may encounter, but there's also the ability for the driver implementation to include extra human-readable information that can be printed out by the user, such as "response timeout" vs "data timeout", or disambiguating "data start" and "data end", etc?

Reply:
Yeah, I like the idea. It's unfortunate that this
```rust
return Err(ErrorKind::Response)
```
needs to become something like this then
```rust
return Err(Error { kind: ErrorKind::Response, message: None })
```
or
```rust
return Err(Error::new(ErrorKind::Response))
```
in that case, although I guess the last option isn't too bad.